### PR TITLE
Overall API Usage improvements to select APIs

### LINKED ISSUE #842: Some charts need client side filtering support
none of the charts has a way to filter data. We need to provide search functionality in some graphs to enhance user experience.

- OVERALL API USAGE ( this needs a select API option same as in 2.6)
- API VERSION USAGE SUMMARY
- API RESOURCE USAGE SUMMARY

### PR DESCRIPTION
**Pull request by fazlan-nazeem**
fixes #842 

Screenshot 2019-12-16 at 11 11 59 AM

### CODE REVIEW

File: components/org.wso2.analytics.apim.widgets/APIMOverallApiUsage/src/APIMOverallApiUsage.jsx [hunk 1]
```diff
@@ -104,6 +104,8 @@ export default function APIMOverallApiUsage(props) {
         },
     };
 
+
+
```
Comment (TanyaM) [inline components/org.wso2.analytics.apim.widgets/APIMOverallApiUsage/src/APIMOverallApiUsage.jsx]: Do we need these new lines?

File: components/org.wso2.analytics.apim.widgets/APIMOverallApiUsage/src/APIMOverallApiUsageWidget.jsx [hunk 2]
```diff
@@ -488,14 +500,44 @@ class APIMOverallApiUsageWidget extends Widget {
...
     }
 
+    selectedAPIChangeCallback = (selectedAPI) => {
+
+        let usageData = this.state.usageData1;
+        let apiIdMapTemp = this.state.apiIdMap;
+        const apiIdMapGlobal = this.state.apiIdMapGlobal;
+        let found = false;
+        const keys = Object.keys(apiIdMapTemp);
+        for(var i in keys) {
```
Comment (TanyaM) [inline components/org.wso2.analytics.apim.widgets/APIMOverallApiUsage/src/APIMOverallApiUsageWidget.jsx]: Fomatting issues. Please run es-lint

File: components/org.wso2.analytics.apim.widgets/APIMOverallApiUsage/src/APIMOverallApiUsageWidget.jsx [hunk 3]
```diff
@@ -488,14 +500,44 @@ class APIMOverallApiUsageWidget extends Widget {
...
+            if(apiIdMapTemp[keys[i]].apiname === selectedAPI[0] && apiIdMapTemp[keys[i]].creator === selectedAPI[1]) {
```
Comment (TanyaM) [inline components/org.wso2.analytics.apim.widgets/APIMOverallApiUsage/src/APIMOverallApiUsageWidget.jsx]: formatting issues.

File: components/org.wso2.analytics.apim.widgets/APIMOverallApiUsage/src/APIMOverallApiUsageWidget.jsx [hunk 4]
```diff
@@ -488,14 +500,44 @@ class APIMOverallApiUsageWidget extends Widget {
...
+                found = true;
+
+                usageData.splice(usageData.findIndex(e => (e[0] === selectedAPI[0]) && (e[1] === selectedAPI[1])), 1);
+            }
+        }
+        if(!found) {
+            const keysglobal = Object.keys(apiIdMapGlobal);
+            let idOfApi= 0;
+            for(var i in keysglobal) {
+                if(apiIdMapGlobal[keysglobal[i]].apiname === selectedAPI[0] && apiIdMapGlobal[keysglobal[i]].creator === selectedAPI[1]) {
```
Comment (TanyaM) [inline components/org.wso2.analytics.apim.widgets/APIMOverallApiUsage/src/APIMOverallApiUsageWidget.jsx]: Exceeds 120 line break?

File: components/org.wso2.analytics.apim.widgets/APIMOverallApiUsage/src/APIMOverallApiUsageWidget.jsx [hunk 5]
```diff
@@ -488,14 +500,44 @@ class APIMOverallApiUsageWidget extends Widget {
...
+                delete apiIdMapTemp[keys[i]];
+                found = true;
+
+                usageData.splice(usageData.findIndex(e => (e[0] === selectedAPI[0]) && (e[1] === selectedAPI[1])), 1);
+            }
+        }
+        if(!found) {
+            const keysglobal = Object.keys(apiIdMapGlobal);
+            let idOfApi= 0;
+            for(var i in keysglobal) {
```
Comment (TanyaM) [inline components/org.wso2.analytics.apim.widgets/APIMOverallApiUsage/src/APIMOverallApiUsageWidget.jsx]: Fomatting issues

File: components/org.wso2.analytics.apim.widgets/APIMOverallApiUsage/src/CustomTable.jsx [hunk 6]
```diff
@@ -187,6 +193,28 @@ class CustomTable extends React.Component {
         this.setState({ query: event.target.value });
     };
 
+    handleSelectedAPIChange = (event) => {
+        let selectedAPIsList = this.state.selectedAPIs;
+        const tickedApi = event.target.value;
+        if(selectedAPIsList.includes(tickedApi)) {
+            for( var i = 0; i < selectedAPIsList.length; i++){
```
Comment (TanyaM) [inline components/org.wso2.analytics.apim.widgets/APIMOverallApiUsage/src/CustomTable.jsx]: Shall we use array functions over loops? WDYT?

File: components/org.wso2.analytics.apim.widgets/APIMOverallApiUsage/src/CustomTable.jsx [hunk 7]
```diff
@@ -241,17 +274,21 @@ class CustomTable extends React.Component {
...
+                                .map((option) => {
                                     return (
                                         <TableRow
                                             hover
                                             tabIndex={-1}
                                         >
                                             <TableCell component='th' scope='row'>
-                                                {n.apiname}
+                                                <Checkbox value={option.apiname}
+                                                                   onChange={this.handleSelectedAPIChange}
```
Comment (TanyaM) [inline components/org.wso2.analytics.apim.widgets/APIMOverallApiUsage/src/CustomTable.jsx]: Formatting issue

File: components/org.wso2.analytics.apim.widgets/APIMOverallApiUsage/src/CustomTable.jsx [hunk 8]
```diff
@@ -264,6 +301,7 @@ class CustomTable extends React.Component {
                         </TableBody>
                     </Table>
                 </div>
+
```
Comment (TanyaM) [inline components/org.wso2.analytics.apim.widgets/APIMOverallApiUsage/src/CustomTable.jsx]: Unnecessary new line

**Comment by fazlan-nazeem** [comment]
@TanyaM I've done the suggested modifications.

File: components/org.wso2.analytics.apim.widgets/APIMOverallApiUsage/src/APIMOverallApiUsage.jsx [hunk 9]
```diff
@@ -157,53 +157,77 @@ export default function APIMOverallApiUsage(props) {
                         />
                     </form>
                 </div>
-                { inProgress ? (
+                {inProgress ? (
```
Comment (ruwiniwj) [inline components/org.wso2.analytics.apim.widgets/APIMOverallApiUsage/src/APIMOverallApiUsage.jsx]: Add a space before `inprogress`

Comment (fazlan-nazeem) [reply components/org.wso2.analytics.apim.widgets/APIMOverallApiUsage/src/APIMOverallApiUsage.jsx]: es-lint is not showing this as a formatting issue. Do we still need to fix?

Comment (ruwiniwj) [reply components/org.wso2.analytics.apim.widgets/APIMOverallApiUsage/src/APIMOverallApiUsage.jsx]: Then no need to fix.

**Pull request merged**